File: apps/admin-v2/src/infrastructure/api/repositories/api-super-admin.repository.ts
import type { SuperAdminRepository, SuperAdminStats } from '@/domain/repositories/super-admin.repository';
import type { Tenant } from '@/domain/entities/tenant';
import type { User } from '@/domain/entities/user';
import type { PaginatedResult } from '@/shared/types/api';
import api from '../client';
import { mapTenant } from '../mappers/tenant.mapper';
import { mapUser } from '../mappers/user.mapper';
import { mapPaginatedResponse } from '../mappers/pagination';

function mapStats(raw: Record<string, unknown>): SuperAdminStats {
  return {
    totalTenants: (raw.total_tenants ?? raw.totalTenants) as number,
    activeTenants: (raw.active_tenants ?? raw.activeTenants) as number,
    totalUsers: (raw.total_users ?? raw.totalUsers) as number,
    totalReservations: (raw.total_reservations ?? raw.totalReservations) as number,
    totalServices: (raw.total_services ?? raw.totalServices) as number,
  };
}

export class ApiSuperAdminRepository implements SuperAdminRepository {
  async getStats(): Promise<SuperAdminStats> {
    const { data } = await api.get('/superadmin/stats');
    return mapStats(data.data ?? data);
  }

  async getTenants(page?: number): Promise<PaginatedResult<Tenant>> {
    const { data } = await api.get('/superadmin/tenants', { params: { page } });
    return mapPaginatedResponse(data, mapTenant);
  }

  async suspendTenant(id: string): Promise<Tenant> {
    const { data } = await api.patch(`/superadmin/tenants/${id}/suspend`);
    return mapTenant(data.data ?? data);
  }

  async activateTenant(id: string): Promise<Tenant> {
    const { data } = await api.patch(`/superadmin/tenants/${id}/activate`);
    return mapTenant(data.data ?? data);
  }

  async getUsers(page?: number): Promise<PaginatedResult<User>> {
    const { data } = await api.get('/superadmin/users', { params: { page } });
    return mapPaginatedResponse(data, mapUser);
  }
}

File: apps/admin_mobile/lib/infrastructure/api/repositories/api_super_admin_repository.dart
import '../../../domain/entities/tenant.dart';
import '../../../domain/entities/user.dart';
import '../../../domain/repositories/super_admin_repository.dart';
import '../../../shared/types/paginated_result.dart';
import '../dio_client.dart';
import '../mappers/tenant_mapper.dart';
import '../mappers/user_mapper.dart';
import 'paginated_helper.dart';

class ApiSuperAdminRepository implements SuperAdminRepository {
  final DioClient _client;

  ApiSuperAdminRepository(this._client);

  @override
  Future<Map<String, dynamic>> getStats() async {
    final response = await _client.dio.get('/superadmin/stats');
    return (response.data['data'] ?? response.data) as Map<String, dynamic>;
  }

  @override
  Future<PaginatedResult<Tenant>> getTenants({int? page}) async {
    final params = <String, dynamic>{};
    if (page != null) params['page'] = page;

    final response =
        await _client.dio.get('/superadmin/tenants', queryParameters: params);
    return extractPaginated(response.data, mapTenant);
  }

  @override
  Future<Tenant> suspendTenant(int id) async {
    final response =
        await _client.dio.patch('/superadmin/tenants/$id/suspend');
    final data = response.data['data'] ?? response.data;
    return mapTenant(data as Map<String, dynamic>);
  }

  @override
  Future<Tenant> activateTenant(int id) async {
    final response =
        await _client.dio.patch('/superadmin/tenants/$id/activate');
    final data = response.data['data'] ?? response.data;
    return mapTenant(data as Map<String, dynamic>);
  }

  @override
  Future<PaginatedResult<User>> getUsers({int? page}) async {
    final params = <String, dynamic>{};
    if (page != null) params['page'] = page;

    final response =
        await _client.dio.get('/superadmin/users', queryParameters: params);
    return extractPaginated(response.data, mapUser);
  }
}
